crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -48,430 +218,376 @@ where
     }
 }
 
-impl<Client, Consensus> BodyDownloader for ConcurrentDownloader<Client, Consensus>
+impl<B, C, DB> BodyDownloader for ConcurrentDownloader<B, C, DB>
 where
-    Client: BodiesClient,
-    Consensus: ConsensusTrait,
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+    DB: Database,
 {
-    fn bodies_stream<'a, 'b, I>(&'a self, headers: I) -> DownloadStream<'a, BlockResponse>
-    where
-        I: IntoIterator<Item = &'b SealedHeader>,
-        <I as IntoIterator>::IntoIter: Send + 'b,
-        'b: 'a,
-    {
-        Box::pin(
-            stream::iter(headers.into_iter())
-                .chunks(self.batch_size)
-                .map(move |headers| {
-                    (move || self.fetch_bodies(headers.clone()))
-                        // We should not backoff from requests since the downloader does
-                        // not know what peer the request is being delivered to.
-                        // See: https://github.com/paradigmxyz/reth/issues/809
-                        .retry(
-                            ConstantBackoff::default()
-                                .with_delay(Duration::from_secs(0))
-                                .with_max_times(self.retries),
-                        )
-                })
-                .buffered(self.concurrency)
-                .map_ok(|blocks| stream::iter(blocks.into_iter()).map(Ok))
-                .try_flatten(),
-        )
+    /// Set a new download range (exclusive).
+    ///
+    /// This method will drain all queued bodies and re-request any
+    /// missing blocks with number less than the last queued block.
+    fn set_download_range(&mut self, range: Range<BlockNumber>) -> Result<(), db::Error> {
+        if range.is_empty() {
+            tracing::warn!(target: "downloaders::bodies", "New header range is empty");
+            return Ok(())
+        }
+
+        tracing::trace!(target: "downloaders::bodies", ?range, "Setting new header range");
+
+        // Drain queued bodies.
+        let queued_bodies = std::mem::take(&mut self.queued_bodies)
+            .into_iter()
+            .filter(|b| range.contains(&b.block_number()))
+            .collect::<Vec<_>>();
+        tracing::trace!(target: "downloaders::bodies", len = queued_bodies.len(), "Evicted bodies from the queue");
+
+        // Dispatch requests for missing bodies
+        if let Some(latest_queued) = self.latest_queued_block_number {
+            if range.start <= latest_queued {
+                let queued_bodies_range = match (queued_bodies.first(), queued_bodies.last()) {
+                    (Some(first), Some(last)) => first.block_number()..last.block_number() + 1,
+                    _ => Range::default(),
```

Comment:
are you doing this so that the loop terminates immediately, instead of doing it with a conditional? "if range_is_empty"?

Reply:
this doesn't have anything to do with loop termination. this is rather an optimization for checking whether the block has already been queued. since we know that queued_bodies is a contiguous sorted collection of blocks, we can construct and check the range as opposed to iterating the array each time checking if the block is there